SI-5905 Sanity check -language options

som-snytt opened the pull request:
The option names are hardcoded, but checked by a test.

There are no hooks to verify options after the compiler
is constructed.

Introduced a `MultiChoiceSetting` required for the
setting creation framework.

som-snytt committed:
SI-5905 Sanity check -language options

The option names are hardcoded, but checked by a test.

There are no hooks to verify options after the compiler
is constructed.

Introduced a `MultiChoiceSetting` required for the
setting creation framework.

som-snytt commented:
Closed until there's way to propose a PR that's not RC.

som-snytt committed:
SI-5905 Restore -language:_

Underscore means all.

-x:c,b,a,_ results in value c,b,a,a,b,c,d,...

Currently, -Xprint does not present phases as a closed
set of choices; there is ad hoc checking in Global.
That would be a nice unification. (You don't know the
list of choices until after global is constructed.)

som-snytt commented:
Reopening for tweaked underscore handling.

som-snytt commented:
Suspended animation.

som-snytt commented:
Reanimating.

som-snytt commented:
I see from another PR that there's an on-hold label used to park PRs?  That would be a PRK.

som-snytt commented:
@adriaanm I know it's not your job anymore, but should I ping someone before I have to rebase?  Pretty soon it will be summer vacation and all of Europe will be hiking in the Alps.

adriaanm commented:
I'm back -- in the Olde Country safe haven. Taking a few more days to adapt to the air, but then I'll be back to terrorizing the PR queue. @retronym is tending to our branches. I believe this one is good to go for review & inclusion in 2.11.1.

retronym commented on test/files/run/t5905-features.scala:
```diff
@@ -0,0 +1,29 @@
+
+// verify that all languageFeature names are accepted by -language
+object Test extends DirectTest {
+  override def code = "class Code { def f = (1 to 10) size }" // exercise a feature
```
I don't understand the need for this.

The following variation of this test passes without this.

``` scala
import tools.partest.DirectTest

// verify that all languageFeature names are accepted by -language
object Test extends DirectTest {
  override def code = ""

  override def extraSettings = s"-usejavacp -d ${testOutput.path}"

  override def show() = {
    val global = newCompiler("")
    new global.Run()
    import global._, global.definitions._

    exitingTyper {
      def isFeature(s: Symbol) = s hasAnnotation LanguageFeatureAnnot
      val langf  = languageFeatureModule.typeSignature
      val feats  = langf.declarations filter (s => isFeature(s)) map (_.name.decoded)
      val xmen   = langf.member(TermName("experimental")).typeSignature.declarations filter (s =>
         isFeature(s)) map (s => s"experimental.${s.name.decoded}"
      )
      val all    = (feats ++ xmen) mkString ","

      assert(feats.nonEmpty, "Test must find feature flags.")

      //dynamics,postfixOps,reflectiveCalls,implicitConversions,higherKinds,existentials,experimental.macros
      compile(s"-language:$all")
    }
  }
}
```

som-snytt replied:
Searching my memory with async i/o, so while I'm waiting for that result, my guess is that I wanted it to emit a warning if the feature flag was wrongly omitted from the test.  Properly, it would exercise all features and ensure no warnings, including no warning about bad flags.

No wait, I didn't read "after the jump."  I'll check on it and amend with corrections and explanatory comments.

Rhetorical edit: was I just being paranoid about global init?

som-snytt committed:
SI-5905 Clarify test case

The language feature options are discovered reflectively, but it
is nice to enforce that expected options are supplied.

Short of that, the code string includes a rowdy postfix operator.

It still does enforce that at least one option was discovered.

Delete -nowarn flags file. Let's see if that was to suppress
a warning in the standard build.

retronym commented:
Can I buy an "S" for the title of the latest commit?

som-snytt commented:
orry about that.

Grabbing text with the mouse in a terminal window is often off by one for me.  I can't get the hang of these newfangled UIs.

adriaanm commented:
LGTM

retronym commented:
I've submitted a PR to revert the last commit here: https://github.com/scala/scala/pull/3732/files

som-snytt commented:
OK, shall I submit the corrected commit against master or 2.11?  I didn't notice it was merged...

retronym commented:
I just submitted a reversion against master. That will be [merged to 2.11.x](https://github.com/scala/scala/pull/3734) in an hour or two. You should target the re-refinement against 2.11.x.
